core/src/main/scala/chisel3/reflect/DataMirror.scala

```diff
@@ -212,6 +212,16 @@ object DataMirror {
     def chiselTypeClone[T <: Data](target: T): T = {
       target.cloneTypeFull
     }
+
+    /** Returns the IOs of a module.
+      *
+      * This method does not necessarily return the final ports of the target module. It consults Chisel's internal data
+      * structures to extract the module's IOs. For this reason, it is generally safer to prefer [[modulePorts]], but
+      * this method may be used for certain use cases that want the current list of ports before the module is closed.
+      *
+      * @param target BaseModule to get IOs from
+      */
+    def moduleIOs(target: BaseModule): Seq[Data] = target.getIOs
```

Comment:
Maybe constraint target to be `FixedIOModule`? I'm afraid in some complex design pattern, this function will execute to early.

Reply:
For the use case I had in mind, I don't want to add that constraint. Indeed, this is not generally safe, which I tried to explain in the scaladoc and by putting this into `DataMirror.internal` object.

Reply:
we need this for more than FixedIOModule.

Reply:
I see, this sounds to be an advance unsafe API, wish normal user won't touch it;p

Reply:
Maybe we can try to add a hook in builder, if user called this function and new IO is created, log an warning to user saying the IO they got is not enough.

Reply:
I think it's fine to start with the new API "hidden" in `DataMirror.internal`, with a comment explaining the dangers. If users start using this and continue to be surprised by the documented behavior, I guess we could add the extra state to give warnings.